### PR TITLE
Restore previously deleted methods associated with MockUp classes.

### PR DESCRIPTION
**Pull request by yukkes**
Thanks you for your excellent work, I have restored previously deleted methods associated with MockUp classes.
I have attempted to fully restore the processing of past versions of JMockit, but have had some setbacks.
I hope you recognize this as an area for future improvement.

* The number of invocations is ignored because there is no judgment process such as @Mock(invocations = 1).

* **The following errors have been resolved. (12/12/2023)**
  * ~~The javadoc generation fails because ConstructorReflection refers to "sun.reflect.*" of a private package in Java 9 or later. The error has been avoided by adding "--ignore-source-errors".~~
  * ~~The test fails because only the last value set in the following test case can be retrieved.~~
  
  
  ```java
     @Test
     public void multiplePublicMockUps(){
        AClass mock1 = new AClassMockUp("Abc").getMockInstance();
        AClass mock2 = new AClassMockUp("Xpto").getMockInstance();
  
        assertNotSame(mock1, mock2);
        assertEquals("Abc", mock1.getTextValue());
        assertEquals("Xpto", mock2.getTextValue());
        assertTrue(AClass.doSomething());
     }
  ```

### CODE REVIEW

**Comment by hazendaz** [comment]
Hi @yukkes, yes very much appreciate this.  The original author I could never understand why he consistently deleted support making upgrades next to impossible.  So I get everything is a work in progress and welcome all your efforts here to restore any needed features that will make your own migrations and others much easier.  I think we all know this library is just maintenance but it needs to work and easily and hopefully up through higher jdks than current as the cost of migrating is high.  I do know open rewrite is working on trying to have a migration path to mockito and are very well aware of this fork.  I think their hurdles have been much of the lost support which makes it hard for them to have the latest here and somehow upgrade the older copies.  So I think this will greatly help so keep it coming :)

**Comment by hazendaz** [comment]
As noted on the other PR, I think we need both 11 and 17 working here before we merge.  21 is going to fail.  I can just remove it from the builds when I get some time as its an ongoing issue.  But the master currently work son 11 and 17.

**Comment by hazendaz** [comment]
For the sun.reflect, can you write that any differently so it might work on 17?  Presume that was when it was either entirely blocked or maybe removed, not sure.  Would need to look.  Its ok if the code doesn't come back exactly like it was.  The main thing is that it just works for the folks that need it.

**Comment by hazendaz** [comment]
holding off this one for now since jdk 17 failed.  Please look into that a bit more.

**Comment by yukkes** [comment]
Thanks for the code review. I'll check back later as I'm spending time with my family right now.

**Comment by yukkes** [comment]
Fixed not to reference the sun.reflect package.
The following blog in Japanese was used as a reference.
https://matsudamper.hatenablog.com/entry/2020/05/17/103231

```java
    public static <T> T newUninitializedInstance(@Nonnull Class<T> aClass) {
        SunReflectionFactoryInstantiator<T> ref = new SunReflectionFactoryInstantiator<>(aClass);
        return ref.newInstance();
    }
```

**Comment by yukkes** [comment]
@hazendaz, thanks for the accurate review remarks.
We have been using JMockit 1.20 for our internal system development and the latest version (1.49) has removed existing functionality, making it difficult to upgrade.
We have now decided to update from Java 8 to 17, and the large amount of test code has made it difficult to migrate to anything but JMockit. So I made JMockit work without modifying the existing test code.

I have decided to commit to this project because we will continue to use JMockit after Java 21 in our company. I hope others will find it useful in the future.

**Comment by yukkes** [comment]
The reason for the test failure in Java 21 is that the class to be mocked is now a sealed class.
The only solution seems to be to ignore the test.

- JDK11, 17
```java
public class InetAddress implements java.io.Serializable {
```

- JDK21
```java
public sealed class InetAddress implements Serializable permits Inet4Address, Inet6Address {
```

**Comment by hazendaz** [comment]
> The reason for the test failure in Java 21 is that the class to be mocked is now a sealed class. The only solution seems to be to ignore the test.
> 
>     * JDK18 or  earlier
> 
> 
> ```java
> public class InetAddress implements java.io.Serializable {
> ```
> 
>     * JDK19 or later
> 
> 
> ```java
> public sealed class InetAddress implements Serializable permits Inet4Address, Inet6Address {
> ```

Agreed, I started looking at migrating most of the tests to junit 5 only using open rewrite but it had a very hard issue here.  So my plan is to do that then simply disable for higher jdk.  Thank you for finding the info.  Do note this did not prevent this from working on java 21.  I use it at work without issues already at that level.  But for our audience faster we fix to show it works on each the more contributors may want to engage and feel more comfortable its working.

File: main/pom.xml
Comment (hazendaz) [inline main/pom.xml:12]: I'll add this back, its a strong preference of mine to be strict in poms of the GAV.  I know m2e in eclipse is all messed up since it has warnings even if set to not warn in recent releases.  Won't bother you with that here, I'll take care of no issues.

File: main/src/main/java/mockit/internal/faking/FakedClassModifier.java
Comment (hazendaz) [inline main/src/main/java/mockit/internal/faking/FakedClassModifier.java:120]: @yukkes Can you add a simple further note here that we are keeping commented out code from original changes?  Normally I'd say delete it but I think in this case I agree commenting it out is fine but want to make sure its obvious why we are keeping commented out code, otherwise someone will delete it it in the future.

File: main/src/test/java/mockit/MisusedFakingAPITest.java
Comment (hazendaz) [inline main/src/test/java/mockit/MisusedFakingAPITest.java:98]: Same on these, just a bit more clarlity as to why we are keeping commented out code.

**Comment by hazendaz** [comment]
@yukkes Very minor asks to make it a bit more clear why intention to keep commented out code.  You may want to date it as well  I typically will add a todo as well so its easily seen, something like this with XXX (your initials)

// TODO 12/2/2023 XXX Reason why we are doing something odd

**Comment by yukkes** [comment]
Please wait for the merge as we are in the process of resolving JMockit-related errors in our internal project.
It is expected to take about a week.

**Comment by hazendaz** [comment]
No problem.  Thank you for all you are doing.

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by yukkes** [comment]
I have restored MockUp class and test class based on JMockit ver. 1.35, please review and merge if there is no problem.

The following test class is restored and all tests PASS.
- MockInvocationProceedTest.java
- MockInvocationTest.java
- MockLoginContextTest.java
- MockUpForGenericsTest.java
- MockUpForSingleClassInstanceTest.java
- MockUpForSingleInterfaceInstanceTest.java

However, the MockUpForSingleInterfaceInstanceTest failed the test as is, so I commented it out with a TODO comment. The reason for this is that the FieldInjection class has changed and no longer sets default values.

**Review by hazendaz — APPROVED** [review]

**Comment by hazendaz** [comment]
@yukkes Are you good at this point for me to push a release?

**Comment by yukkes** [comment]
@hazendaz Yes, please release.
The changes include restoring the MockUp and Deencapsulation classes based on JMockit ver. 1.35.

Note that the annotation to determine the number of method invocations has been restored as shown below, but since there is no judgment process, there is no error even if the number of invocations is different.
`@Mock(invocations=1)`

**Comment by hazendaz** [comment]
Release is out now.  Let me know how that works out for you.

**Comment by yukkes** [comment]
@hazendaz Thank you for the release.
I will check if the unit test passes using the latest version on our own project.

**Comment by yukkes** [comment]
@hazendaz
I have performed my own unit tests and the results match my own build and ver. 1.52.0, so it looks fine. Thanks.

**Pull request merged**
